### ISSUE TITLE
Builder pattern using Tags leads to high CPU overhead

### ISSUE THREAD

**Issue by DanielThomas** [micrometer-metrics/micrometer#4052]
We discovered some significant CPU hotspots in a service and found several places in our frameworks that were doing a significant amount of unnecessary work due to the  sort/dedup on every `Tags` class operation. In our case we were either using `Tags` as a builder with up to a dozen calls to `Tags::and` or has several calls to `tags(String, String)` on the builders.

Appears when this commit landed it switched all builders to use`Tags` internally, so incur the sort/dedup penalty for every call, making building tags quite expensive:

https://github.com/micrometer-metrics/micrometer/commit/b7bbea3d6935aaf794d538cf5f6da5155c9bb93f#diff-a64eec4db3aaf5a50e01e46efed55b3eecb772bcbe0ad929ef59b455c7a5ec2c

This makes `Tags` a poor choice for intermediate `Tags` state and if sort/dedupe can't be avoided, there should probably be a mutable builder class backing these usages instead. If that overhead can't be avoided, I'd argue that the builder-style methods on `Tags` should be deprecated, because it's easy to get into trouble and not realise it.

The starting state for the builders is also `Tags.EMPTY` so `Tags::and` could probably use an optimization tha returns the incoming `Tags` instance if you're calling `and` against `Tags.EMPTY`.

**Environment**

 - Micrometer version 1.11.3

**To Reproduce**

Use multiple calls to `tags` 

**Expected behavior**

Tag construction incurs as little overhead as possible.

**Additional context**

https://github.com/Netflix/dgs-framework/pull/1612

**Comment by DanielThomas** [comment]
I should add, if there's the possibility of avoiding this completely, that'd be great - before these changes, these accounted for 50%(!) of the CPU overhead for the application. We've gotten that down a bunch, but it's unavoidable when doing the final measurement:

<img width="1145" alt="Screenshot 2023-09-08 at 6 10 51 pm" src="https://github.com/micrometer-metrics/micrometer/assets/1479220/13e10ce6-5af3-490b-9a74-fd090217fc8c">

**Comment by shakuzen** [comment]
Thank you for the report and details. It's much appreciated. We would need to look into this carefully. Unfortunately, I think the complete motivation/context in which the original changes were made (almost 5 years ago, for version 1.1.0) is lost at this point. So we'd really need to dig into all things that could be affected by making changes around this.

> I should add, if there's the possibility of avoiding this completely, that'd be great

To be clear, you mean avoiding the sort and deduplication?

> before these changes, these accounted for 50%(!) of the CPU overhead for the application.

That's unexpected. There is certainly more overhead with it being immutable and sorting/deduping each call to `and`, but it should still be a rather small overhead in absolute terms. Is the application just not doing much work on the CPU?

**Comment by DanielThomas** [comment]
Yes, particularly the sort. The deduplication isn't so much of a concern, but still feels like work that'd be better to avoid until the last possible moment.

This particular application is a DGS framework application, which has timers around GraphQL field data fetchers, which is what you're seeing in that last screenshot. Those data fetchers also use Spring Security method authorization and we have counters for authz decisions too, so each incoming query will have dozens of metrics recorded.

This service mostly spends most of the time in GRPC doing the data fetching to other queries, which is why instrumentation can be such a CPU hotspot compared to the other work going on.

**Comment by jiangxinlingdu** [comment]
```
    private Tags(Tag[] tags) {
        this.tags = tags;
        Arrays.sort(this.tags);
        dedup();
    }
```
i think sort unnecessary
